sqlglot/transforms.py

```diff
@@ -398,6 +398,41 @@ def eliminate_full_outer_join(expression: exp.Expression) -> exp.Expression:
     return expression
 
 
+def move_ctes_to_top_level(expression: exp.Expression) -> exp.Expression:
+    """
+    Some dialects (e.g. Hive, T-SQL, Spark prior to version 3) only allow CTEs to be
+    defined at the top-level, so for example queries like:
+
+        SELECT * FROM (WITH t(c) AS (SELECT 1) SELECT * FROM t) AS subq
+
+    are invalid in those dialects. This transformation ensures all CTEs are moved
+    to the top level so that the final SQL code is valid from a syntax standpoint.
+
+    TODO: handle name clashes whilst moving CTEs (it can get quite tricky & costly).
+    """
+    # This transformation is only applied to the root expression, if it supports CTEs
+    if expression.parent or "with" not in expression.arg_types:
+        return expression
+
+    expression = expression.copy()
```

Comment:
transforms shouldn't manually copy, this should be done by the caller

Reply:
if you want to be most efficient, you could only copy if you find With Statements

Reply:
Agree, this transformation was initially defined in dialect.py and so the copy was left over since it wasnt supposed to be used in combination with preprocess.

Reply:
Addressed both comments